userspace/libsinsp/sinsp_filtercheck_thread.cpp

```diff
@@ -761,21 +825,104 @@ uint8_t* sinsp_filter_check_thread::extract(sinsp_evt *evt, OUT uint32_t* len, b
 	case TYPE_ENV:
 		{
 			m_tstr.clear();
-
 			uint32_t j;
 			const auto& env = tinfo->get_env();
 			uint32_t nargs = (uint32_t)env.size();
 
-			for(j = 0; j < nargs; j++)
+			if(nargs == 0)
 			{
-				m_tstr += env[j];
-				if(j < nargs -1)
+				RETURN_EXTRACT_STRING(m_tstr);
+			}
+
+			if(!m_argname.empty()) // contains ENV_NAME we are looking to extract
+			{
+				// proc.env[ENV_NAME] use case: returns matched env variable value
+				for (const auto& item : env)
 				{
-					m_tstr += ' ';
+					std::string str(item.data(), item.data() + item.size());
+					if(str.find(m_argname) == 0)
+					{
+						size_t pos = str.find('=');
+						if (pos != std::string::npos)
+						{
+							m_tstr = str.substr(pos + 1);
+							RETURN_EXTRACT_STRING(m_tstr);
+						}
+					}
+				}
+				RETURN_EXTRACT_STRING(m_tstr); // return empty in case of no match
+			} else
+			{ // proc.env return concatenated env string of format "ENV_NAME=value ENV_NAME1=value1" ...
+				for(j = 0; j < nargs; j++)
+				{
+					m_tstr += env[j];
+					if(j < nargs -1)
+					{
+						m_tstr += ' ';
+					}
 				}
+				RETURN_EXTRACT_STRING(m_tstr);
 			}
+		}
+	case TYPE_AENV:
+		{
+			m_tstr.clear();
+			uint32_t j;
+			const auto& env = tinfo->get_env();
+			uint32_t nargs = (uint32_t)env.size();
 
-			RETURN_EXTRACT_STRING(m_tstr);
+			if(!m_argname.empty()) // contains ENV_NAME we are looking to extract
+			{
+				sinsp_threadinfo* mt = NULL;
+				if(tinfo->is_main_thread())
+				{
+					mt = tinfo;
+				}
+				else
+				{
+					mt = tinfo->get_main_thread();
+					if(mt == NULL)
+					{
+						RETURN_EXTRACT_STRING(m_tstr);
+					}
+				}
+
+				for(int32_t j = 0; j < 7; j++) // up to 7 levels
```

Comment:
I think we can also go up to 20/30 levels just to not exclude corner cases, I expect that in almost all cases we will break the loop after a few iterations because we face `if(mt == NULL)`. Let's say that we need to put an upper limit just to avoid possible misbehaviors but i don't expect to have process in the system with 20 ancestors :exploding_head:

Reply:
haha 20 it is!